manifests/install.pp

```diff
@@ -15,21 +15,47 @@
 
   case $consul::install_method {
     'url': {
+      # This is done just so we can take a look at what staging path it wants to use.
+      # If we remove this, we have to completely assume or manage the staging path
+      include staging
+
       if $::operatingsystem != 'darwin' {
-        ensure_packages(['unzip'], { 'before' => Staging::File['consul.zip'] })
+        ensure_packages(['unzip'], { 'before' => Staging::File[$consul::real_download_file] })
+      }
+
+      ## Suggest moving to using puppetlabs-transition module to do the service work
+      ##  after it gets some patchwork and closer to release version
+      #
+      # This was done for puppet 3.x not supporting Ubuntu 15 and Fedora 22, and since this ruby line doesn't support site.pp... override of Service { provider => 'systemd' }
+      if $::puppetversion =~ /^4/ {
+        $ruby_service_stop = "ruby -r 'puppet' -e \"Puppet::Type.type(:service).newservice(:name => 'consul').provider.send('stop')\""
+      } else {
+        $ruby_service_stop = $::operatingsystem ? {
+          'fedora' => "ruby -r 'puppet' -e \"Puppet::Type.type(:service).newservice(:name => 'consul', :provider => '${consul::init_style}').provider.send('stop')\"",
+          'ubuntu' => "ruby -r 'puppet' -e \"Puppet::Type.type(:service).newservice(:name => 'consul', :provider => '${consul::init_style}').provider.send('stop')\"",
+          default  => "ruby -r 'puppet' -e \"Puppet::Type.type(:service).newservice(:name => 'consul').provider.send('stop')\""
+        }
+      }
+      # I don't trust mistakes in $staging::path as if it was set to / then this find would delete everything except that one file all the way from the root path /
+      # And since we're only supporting Linux and Darwin(Mac) in this current revision of this puppet-consul module.
+      # This cleans all files in the staging folder except this specific file
+      if $staging::path =~ /^\/opt\/staging.*/ {
+        $clean_staging_dir = "find ${staging::path}/consul ! -name '${consul::real_download_file}' -type f -exec rm -f {} +"
+      } else {
+        $clean_staging_dir = "find /opt/staging/consul ! -name '${consul::real_download_file}' -type f -exec rm -f {} +"
       }
-      staging::file { 'consul.zip':
+      staging::file { $consul::real_download_file:
         source => $consul::real_download_url
       } ->
-      staging::extract { 'consul.zip':
-        target  => $consul::bin_dir,
-        creates => "${consul::bin_dir}/consul",
+      staging::extract { $consul::real_download_file:
+        target => $consul::bin_dir,
+        unless => "which consul > /dev/null ; if [ $? = 0 ]; then test `consul version | grep -m1 -o [0-9\\.] | tr -d '\\n'` = ${consul::version}; unlessval=$?; if [ \$unlessval = 1 ]; then ${clean_staging_dir}; ${ruby_service_stop}; rm -f ${consul::bin_dir}/consul; fi; else unlessval=1; fi; test \$unlessval = 0",
```

Comment:
Yup. When upgrading consul on a server it requires manual intervention. I usually stop puppet on *all* masters and then run puppet on each. I'll write a fix to optionally allow us to toggle service restarts so we can disable them on master so we don't have to keep stopping puppet.